### ISSUE TITLE
Policies not enforced correctly when HostPolicy is enabled for docker version 18.06 and earlier

### ISSUE THREAD

**Issue by oneiro-naut** [kubearmor/KubeArmor#223]
Some policies are not working correctly after hostpolicy is enabled in kubearmor. I think this might be some AppArmor issue.
This issue happens regardless of the kernel version. I tested this on both kubearmor-dev and kubearmor-dev-next.
Docker versions affected: versions older than 18.09 (tested on 18.03 and 18.06)
Steps to reproduce the issue:
CASE I: With HostPolicies enabled
1. Run kubearmor with -enableHostPolicy.
```
vagrant@kubearmor-dev-next:~/KubeArmor/KubeArmor$ make && make run-with-host
```
2. Deploy multiubuntu pods.
3. Apply sample policy [ksp-ubuntu-3-proc-dir-allow.yaml](https://github.com/kubearmor/KubeArmor/blob/master/examples/multiubuntu/security-policies/ksp-ubuntu-3-proc-dir-allow.yaml). (This policy allows binaries in directory /bin to be able to run)
4. Run some binary on the ubuntu-3-deployment pod from the /bin directory.
```
vagrant@kubearmor-dev-next:~$ kubectl exec -ti -n multiubuntu ubuntu-3-deployment-59bf5f5564-b2xxz -- /bin/bash
command terminated with exit code 129
vagrant@kubearmor-dev-next:~$ kubectl exec -n multiubuntu ubuntu-3-deployment-59bf5f5564-b2xxz -- /bin/bash
command terminated with exit code 139
vagrant@kubearmor-dev-next:~$ kubectl exec -n multiubuntu ubuntu-3-deployment-59bf5f5564-b2xxz -- /bin/echo "test"
command terminated with exit code 139
```
Observation I: The policy is not working since we are not able to run binaries in /bin directory. This is unexpected result.
5. Delete multiubuntu deployment.

CASE II: WIthout HostPolicies enabled
1. Run kubearmor without -enableHostPolicy.
```
vagrant@kubearmor-dev-next:~/KubeArmor/KubeArmor$ make && make run
```
2. Deploy multiubuntu pods.
3. Apply sample policy [ksp-ubuntu-3-proc-dir-allow.yaml](https://github.com/kubearmor/KubeArmor/blob/master/examples/multiubuntu/security-policies/ksp-ubuntu-3-proc-dir-allow.yaml). (This policy allows binaries in directory /bin to be able to run)
4. Run some binary on the ubuntu-3-deployment pod from the /bin directory.
```
vagrant@kubearmor-dev-next:~$ kubectl exec -ti -n multiubuntu ubuntu-3-deployment-59bf5f5564-cbcnw -- /bin/bash
bash: /etc/bash.bashrc: Permission denied
bash: /root/.bashrc: Permission denied
bash-4.4# echo "test"
test
bash-4.4# 
```
Observation II: The policy works since we are able to run binaries in /bin directory. This is expected result.

Ideally we should get the same result for both the cases .

Edit: The issue is version specific and does not happen for docker versions >= 18.09.

**Comment by oneiro-naut** [comment]
@nam-jaehyun This doesn't happen with newer docker versions. I tested it with this docker version:
```
 Version:           20.10.8
 API version:       1.41
``` 
on both kernel v4.18 and v5.13. I wasn't able to reproduce the issue with this docker version. It seems to be happening for older docker versions only(like v18.x).

**Comment by nam-jaehyun** [comment]
@oneiro-naut then, does it happen with docker version 19.03? I mean I'd like to make sure that it happens with 18.x only.
Also, I'm wondering how enabling host policies affects container policies according to the docker version.

**Comment by oneiro-naut** [comment]
@nam-jaehyun I tested for versions 19.03 and 18.09.x and it seems to work. So we can say that this does not happen for version 18.09.1 or greater. For versions earlier than this we get this issue. I checked with v18.06 and 18.03(the default docker version in ubuntu 18.04) and the issue is present. Starting with version 18.09 docker daemon stopped managing containerd so it is somehow related to that I guess. But I am not sure about that. Also I noticed kubearmor.host blocks certain signals sent by docker whenever this issue is happening. When this issue is triggered containers cant be deleted either(it gets stuck). This happens each time this issue happens. This is what I get in `/var/log/audit/audit.log`:
```
type=AVC msg=audit(1628499029.432:580): apparmor="DENIED" operation="signal" profile="kubearmor-multiubuntu-ubuntu-2-container" pid=15312 comm="docker-runc" requested_mask="receive" denied_mask="receive" signal=term peer="kubearmor.host"
type=AVC msg=audit(1628499030.449:581): apparmor="ALLOWED" operation="signal" profile="kubearmor.host" pid=15324 comm="docker-runc" requested_mask="send" denied_mask="send" signal=kill peer="kubearmor-multiubuntu-ubuntu-4-container"
type=AVC msg=audit(1628499030.449:581): apparmor="DENIED" operation="signal" profile="kubearmor-multiubuntu-ubuntu-4-container" pid=15324 comm="docker-runc" requested_mask="receive" denied_mask="receive" signal=kill peer="kubearmor.host"
type=AVC msg=audit(1628499031.425:582): apparmor="ALLOWED" operation="signal" profile="kubearmor.host" pid=15332 comm="docker-runc" requested_mask="send" denied_mask="send" signal=term peer="kubearmor-multiubuntu-ubuntu-1-container"
type=AVC msg=audit(1628499031.425:582): apparmor="DENIED" operation="signal" profile="kubearmor-multiubuntu-ubuntu-1-container" pid=15332 comm="docker-runc" requested_mask="receive" denied_mask="receive" signal=term peer="kubearmor.host"
type=AVC msg=audit(1628499031.445:583): apparmor="ALLOWED" operation="signal" profile="kubearmor.host" pid=15340 comm="docker-runc" requested_mask="send" denied_mask="send" signal=kill peer="kubearmor-multiubuntu-ubuntu-2-container"
type=AVC msg=audit(1628499031.445:583): apparmor="DENIED" operation="signal" profile="kubearmor-multiubuntu-ubuntu-2-container" pid=15340 comm="docker-runc" requested_mask="receive" denied_mask="receive" signal=kill peer="kubearmor.host"
```
I think this issue related to signal has to do with https://github.com/moby/moby/pull/37831 (which is fixed for docker version 18.09.1 onwards).

**Comment by oneiro-naut** [comment]
@nam-jaehyun Since this issue and #248 are version specific and seem to be fixed for Docker version >=18.09 there is a need to update the minimum docker version requirement for kubearmor from version 18.03 to 18.09(until these issues are solved). I have raised a PR(#242) which makes this update.
